runtime/cni_integ_test.go

```diff
@@ -0,0 +1,212 @@
+	fccontrol "github.com/firecracker-microvm/firecracker-containerd/proto/service/fccontrol/ttrpc"
+	"github.com/firecracker-microvm/firecracker-containerd/runtime/firecrackeroci"
+)
+
+func TestCNISupport_Isolated(t *testing.T) {
+	internal.RequiresIsolation(t)
+
+	testTimeout := 120 * time.Second
+	ctx, cancel := context.WithTimeout(namespaces.WithNamespace(context.Background(), defaultNamespace), testTimeout)
+	defer cancel()
+
+	client, err := containerd.New(containerdSockPath, containerd.WithDefaultRuntime(firecrackerRuntime))
+	require.NoError(t, err, "unable to create client to containerd service at %s, is containerd running?", containerdSockPath)
+	defer client.Close()
+
+	pluginClient, err := ttrpcutil.NewClient(containerdSockPath + ".ttrpc")
+	require.NoError(t, err, "failed to create ttrpc client")
+
+	image, err := alpineImage(ctx, client, naiveSnapshotterName)
+	require.NoError(t, err, "failed to get alpine image")
+
+	numVMs := 5
+	var vmIDs []string
+	webpages := make(map[string]string)
+	for i := 0; i < numVMs; i++ {
+		vmID := fmt.Sprintf("vm-%d", i)
+		vmIDs = append(vmIDs, vmID)
+		webpages[vmID] = fmt.Sprintf("Hello, my virtual machine %s\n", vmID)
+	}
+
+	localServices, err := internal.NewLocalNetworkServices(webpages)
+	require.NoError(t, err, "failed to create local network test services")
+
+	cniNetworkName := "fcnet-test"
+	err = writeCNIConf("/etc/cni/conf.d/fcnet-test.conflist", cniNetworkName, localServices.DNSServerIP())
+	require.NoError(t, err, "failed to write test cni conf")
+
+	go func() {
+		require.NoError(t, localServices.Serve(ctx), "failed serving local dns and http")
+	}()
+
+	var vmGroup sync.WaitGroup
+	for _, vmID := range vmIDs {
+		vmGroup.Add(1)
+		go func(vmID string) {
+			defer vmGroup.Done()
+
+			fcClient := fccontrol.NewFirecrackerClient(pluginClient.Client())
+			_, err = fcClient.CreateVM(ctx, &proto.CreateVMRequest{
+				VMID: vmID,
+				MachineCfg: &proto.FirecrackerMachineConfiguration{
+					MemSizeMib: 512,
+				},
+				RootDrive: &proto.FirecrackerDrive{
+					PathOnHost:   defaultVMRootfsPath,
+					IsReadOnly:   false,
+					IsRootDevice: true,
+				},
+				NetworkInterfaces: []*proto.FirecrackerNetworkInterface{{
+					CNIConfig: &proto.CNIConfiguration{
+						NetworkName: cniNetworkName,
+						IfName:      "veth0",
+					},
+				}},
+			})
+			require.NoError(t, err, "failed to create vm")
+
+			containerName := fmt.Sprintf("%s-container", vmID)
+			snapshotName := fmt.Sprintf("%s-snapshot", vmID)
+
+			newContainer, err := client.NewContainer(ctx,
+				containerName,
+				containerd.WithSnapshotter(naiveSnapshotterName),
+				containerd.WithNewSnapshot(snapshotName, image),
+				containerd.WithNewSpec(
+					oci.WithProcessArgs("/usr/bin/wget",
+						"-q",      // don't print to stderr unless an error occurs
+						"-O", "-", // write to stdout
+						localServices.URL(vmID)),
+					firecrackeroci.WithVMID(vmID),
+					firecrackeroci.WithVMNetwork,
```

Comment:
How come this is specified here but not in `taskworkflow.go`?

I'm wondering if this is something we really want the client to have to specify or not.  We might want to default to using the VM network when nothing is specified and to allow opting out instead.

Reply:
I think we can have users specify the option for now.  I'm trying to anticipate what it would look like to try and use the existing cri-containerd plugin with firecracker-containerd, and seeing whether we can avoid requiring changes.  Might be worth opening an issue for this so we can follow up later on.

Reply:
> What do you think we should open an issue about? Finding a way to set these options by default?  Or figuring out how this will interact with CRI?

I think "figuring out how this will interact with CRI" is the right answer right now.  It *might* also be reasonable to set these options by default.  Either way, I think it'd be good to move this conversation to an issue.

> My impression was that CRI is making containerd CreateTask calls itself

So I'd like to disambiguate a bit.  CRI is a [spec](https://kubernetes.io/blog/2016/12/container-runtime-interface-cri-in-kubernetes/) with multiple implementations.  [cri-containerd](https://github.com/containerd/cri) is a containerd plugin that implements CRI and is bundled by default into containerd.

> We'd instead want just CRI to create its sandbox container using "host" netns (which would actually be the Firecracker VM's).

Yeah, this is probably a reasonable approach as well.

Reply:
Commented on this issue, can continue discussion there: https://github.com/firecracker-microvm/firecracker-containerd/issues/100